source/module_hamilt_pw/hamilt_pwdft/operator_pw/projop_pw.cpp

```diff
@@ -0,0 +1,136 @@
+    }
+
+    template<typename T, typename Device>
+    DFTU<OperatorPW<T, Device>>::DFTU(const std::vector<int> isk,
+                                      const std::vector<int>& l_hubbard,
+                                      const std::vector<double>& u,
+                                      const std::vector<double>& rgrid,
+                                      const std::vector<std::vector<double>>& projs,
+                                      const std::vector<int>& natom,
+                                      const std::vector<ModuleBase::Vector3<double>*>& tau,
+                                      const double& omega,
+                                      const double& tpiba,
+                                      const std::vector<ModuleBase::Vector3<double>>& q,
+                                      const double& dq,
+                                      const double& nq)
+    {
+        RadialProjection::RadialProjector rp;
+        const int nr = rgrid.size();
+#ifdef __DEBUG
+        for(auto &proj: projs)
+        {
+            assert(proj.size() == nr);
+        }
+#endif
+
+        rp._build_sbt_tab(rgrid,     // std::vector<double>
+                          projs,     // std::vector<std::vector<double>>
+                          l_hubbard, // std::vector<int>
+                          nq,        // int
+                          dq);       // double
+        
+        rp.sbtft(q,             // std::vector<ModuleBase::Vector3<double>>
+                 proj_q_tab_,   // std::vector<std::complex<double>>
+                 'r',           // char, 'r' for <G+k|p>, 'l' for <p|G+k>
+                 omega,         // double
+                 tpiba          // double
+                 );
+    }
+
+    template<typename T, typename Device>
+    DFTU<OperatorPW<T, Device>>::DFTU(const std::vector<int>& isk,
+                                      const UnitCell* ucell_in,
+                                      const ModulePW::PW_Basis_K* pw_basis)
+    {
+        const double omega = ucell_in->omega;
+        const double tpiba = ucell_in->tpiba;
+        std::vector<ModuleBase::Vector3<double>> q;
+        for(int ig = 0; ig < pw_basis->npw; ++ig)
+        {   // the following line is incorrect, because I am not clear about what the isk is!
+            q.push_back(pw_basis->getgpluskcar(isk[0], ig)); // not isk[0], but actually loop on isk
+        }
+        const int ntype = ucell_in->ntype;
+
+        std::vector<int> l_hubbard(ntype);   // read from UnitCell or somewhere else
+        std::vector<double> u(ntype);        // read from UnitCell or somewhere else
+        std::vector<double> onsite_r(ntype); // read from UnitCell or somewhere else
+        
+        std::vector<std::vector<double>> projs(ntype);
+        int nr = -1;
+        double dr;
+        bool padding = false;
+        for(int it = 0; it < ntype; ++it)
+        {
+            std::vector<std::vector<double>> temp_;
+            int lmax, nr_;
```

Comment:
initialize the variables, one line for one variable is better

Reply:
thanks, changes were made in the latest commit.